rpcs3/Emu/RSX/Common/ProgramStateCache.h

```diff
@@ -195,51 +197,60 @@ class program_state_cache
 	/// bool here to inform that the program was preexisting.
 	std::tuple<const vertex_program_type&, bool> search_vertex_program(const RSXVertexProgram& rsx_vp, bool force_load = true)
 	{
-		const auto& I = m_vertex_shader_cache.find(rsx_vp);
-		if (I != m_vertex_shader_cache.end())
+		vertex_program_type* new_shader;
 		{
-			return std::forward_as_tuple(I->second, true);
-		}
+			std::lock_guard lock(m_vertex_mutex);
```

Comment:
Replace these with proper rw locks. Look at rpcs3's shared_mutex class for that.

Reply:
I did give read/writer locks a try, but there seems to be some kind of iterator invalidation going on when a thread writes to the unordered_map after another thread already read from it. Maybe when it has to rehash to fit more data? 

Either that or I'm doing something wrong:

```suggestion
                vertex_program_type* new_shader;
		{
			reader_lock lock(m_vertex_mutex);

			const auto& I = m_vertex_shader_cache.find(rsx_vp);
			if (I != m_vertex_shader_cache.end())
			{
				return std::forward_as_tuple(I->second, true);
			}

			if (!force_load)
			{
				return std::forward_as_tuple(__null_vertex_program, false);
			}

			LOG_NOTICE(RSX, "VP not found in buffer!");
			lock.upgrade();
			new_shader = &m_vertex_shader_cache[rsx_vp];
		}
		backend_traits::recompile_vertex_program(rsx_vp, *new_shader, m_next_id++);
```

Reply:
That's wrong

Reply:
You lock at the begining of the function, the destructor of the lock will unlock automatically. The code is the same but the lock scope should not exist.

Reply:
Can't figure out the suggestion system, but theflow should be:
```
function()
{
    reader_lock lock(shared_mutex);
    map_access_stuff;
    conditional_return;

    lock.upgrade();
    change map contents;
}
```

Reply:
The lock scope is there to release the lock before calling recompile_vertex_program and recompile_fragment_program so multiple threads are able to call them at the same time. 

This way it a good chunk of the speedup is gone and I'm still getting memory corruption:

```suggestion
/// bool here to inform that the program was preexisting.
	std::tuple<const vertex_program_type&, bool> search_vertex_program(const RSXVertexProgram& rsx_vp, bool force_load = true)
	{
		reader_lock lock(m_vertex_mutex);

		const auto& I = m_vertex_shader_cache.find(rsx_vp);
		if (I != m_vertex_shader_cache.end())
		{
			return std::forward_as_tuple(I->second, true);
		}

		if (!force_load)
		{
			return std::forward_as_tuple(__null_vertex_program, false);
		}

		LOG_NOTICE(RSX, "VP not found in buffer!");
		lock.upgrade();
		vertex_program_type new_shader = m_vertex_shader_cache[rsx_vp];
		backend_traits::recompile_vertex_program(rsx_vp, new_shader, m_next_id++);

		return std::forward_as_tuple(new_shader, false);
	}
```

Reply:
You almost got it right there, you need the scope to end before calling recompile_vertex_program. However, this is a MT cache which means you must recheck before finishing; i.e you have to check if someone else added the vp to the map before assigning unconditionally. That's why you get memory problems.